sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2ScanExec.scala

```diff
@@ -46,34 +48,46 @@ case class DataSourceV2ScanExec(
       new DataSourcePartitioning(
         s.outputPartitioning(), AttributeMap(output.map(a => a -> a.name)))
 
+    case _ if readerFactories.size == 1 => SinglePartition
+
     case _ => super.outputPartitioning
   }
 
-  private lazy val readerFactories: java.util.List[DataReaderFactory[UnsafeRow]] = reader match {
-    case r: SupportsScanUnsafeRow => r.createUnsafeRowReaderFactories()
+  private lazy val readerFactories: Seq[DataReaderFactory[_]] = reader match {
```

Comment:
Why not separate the cases for columnar batch and unsafe rows? That would avoid needing to cast this later to `Seq[DataReaderFactory[UnsafeRow]]` and `Seq[DataReaderFactory[ColumnarBatch]]`, which isn't a very clean solution.

Reply:
I agree that we should separate these cases, but I don't immediately see a good way to do so.

Reply:
Why not separate the method into `batchReaderFactories` and `readerFactories`?

Reply:
We could, but then both usages of `readerFactories` (along with any future ones) would have to add a `case r: SupportsScanColumnarBatch if r.enableBatchRead()` check to figure out which val they're supposed to use. I'm not a huge fan of lazy vals that will throw errors if anyone tries to instantiate them.

Both ways seem equally unclean to me, so I'm fine with switching if you think separating them is better.

Reply:
I think it is better to have a lazy val throw an exception if it is called at the wrong time (if it is private) than to add the cast because the exception at least validates that assumptions are met and can throw a reasonable error message. The cast might hide the problem, particularly over time as this code evolves. It would be reasonable to add another path that returns `Seq[DataReaderFactory[_]]` because that's the method contract, even though there is an assumption in the callers about how it will behave. As much of the contract between methods as possible should be expressed in types.

Reply:
That's a good point. Updated the PR.